WIP: HomeKit sensors - Koogeek

drndos opened the pull request:
Add HomeKit Controller sensor
Fix dispatching mechanism for unknown services

## Description:
This Pull request adds support for custom vendor specific homekit services and specifically Koogeek sensor support. It was tested on Koogeek O1EU homekit device.
![image](https://user-images.githubusercontent.com/5576134/47952856-ecb12480-df75-11e8-8680-4d2b26e45ac1.png)

Ignoring TLV8 encoded data from sensor data for now.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7202

## Example entry for `configuration.yaml` (if applicable):
N/A

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - N/A New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - N/A New dependencies are only imported inside functions that use them ([example][ex-import]).
  - N/A New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

drndos committed:
Add Koogeek sensor platform
Add Dispatching mechanism for vendor specific homekit services to corresponding platforms
Add constant for watt-hours

drndos committed:
Fix line lengths

drndos committed:
Add the koogeek homekit platform to coveragerc

frenck commented:
Could not find a matching PR on our documentation repository. Adding `docs-missing` label.

drndos commented:
HomeKit is using Tlv8 format for some of the values (base64 looking values). So I decided to omit them in the initial release, since they don't bring that much value to end user.

drndos committed:
Remove TLV8 encoded values from initial version of Koogeek Sensor

MartinHjelmare commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -148,11 +156,25 @@ def accessory_setup(self):
+                                  kservice['type'],
+                                  kservice['platform'])
+                    component = kservice['type']
+                    devdomain = kservice['platform']
```
I don't think we want to do this. A component should only set up its own platforms.

MartinHjelmare replied:
This would need approval by core or in architecture issue. But I don't think it will be approved.

balloob replied:
This should not be a standalone platform but instead be part of homekit sensor.

drndos replied:
This is specific implementation for Koogeek service, and not the generic implementation of HomeKit Service that would be handled by HomeKit Controller Sensor. I can do it in HomeKit Controller sensor but it is specific only to Koogeek. After a while the HomeKit Controller Sensor/Switch ... would get too big with all the different vendor's service implementations. I am not sure it would be correct. Do you think it would be better in HomeKit Controller Sensor even if it is vendor specific ?

I tried asking on Discord without luck so I implemented it somehow to get feedback.

balloob replied:
it's fine if the homekit sensor grows too big, we can always split it up in multiple files (`sensor/homekit/koogeek.py` etc)

balloob replied:
CC @cdce8p

cdce8p replied:
CC: @mjg59

drndos replied:
Oki thanks for clarification, will do it that way then :)
What do you think about the "dispatching" (mapping) of HomeKit services to Home Assistant ? Do you think that it should be in HomeKit Controller like now or do you think it would be better in switch/sensor ... ?

drndos committed:
Fix Refactor Koogeek platform to HomeKit Controller Sensor
Fix Change dispatch mechanism of custom HomeKit services
Revert new constant of Watt Hours from constants
Fix .coveragerc inline with refactor

drndos commented:
I have changed the dispatch mechanism so that HomeKit Controller doesn't create different platform.
I have changed the Koogeek sensor to be inside a HomeKit Sensor and implemented generic way to set default unit of measurement and default status of sensor.

drndos committed:
Add support for generic HomeKit sensors

drndos committed:
Fix line length

mjg59 commented on homeassistant/components/homekit_controller/__init__.py:
```diff
@@ -25,6 +25,7 @@
     'lightbulb': 'light',
     'outlet': 'switch',
     'thermostat': 'climate',
+    '4aaaf930-0dec-11e5-b939-0800200c9a66': 'sensor'
```
Ideally this would be added to the upstream homekit library rather than leaving it as a raw UUID here - the main concern is that if you leave it as the raw value and it later gets added, this will probably silently break.

mjg59 replied:
(Or is this UUID Kogeek-specific?)

drndos replied:
There is one issue with generic services and characteristics though in sensor and that is why I didn't enable it without proper testing. Which characteristic is the default one and what is its unit of measurement.

mjg59 reviewed (commented):
Otherwise, looks good

balloob commented on .coveragerc:
```diff
@@ -713,7 +713,7 @@ omit =
     homeassistant/components/sensor/iperf3.py
     homeassistant/components/sensor/irish_rail_transport.py
-    homeassistant/components/sensor/koogeek.py
+    homeassistant/components/sensor/homekit_controller.py
```
🔡

balloob commented on homeassistant/components/sensor/homekit_controller.py:
```diff
@@ -1,54 +1,71 @@
-    '4aaaf932-0dec-11e5-b939-0800200c9a66': 'current_hour_data',
-    '4aaaf93f-0dec-11e5-b939-0800200c9a66': 'running_time'
+CHARACTERISTICS = {
+    '4aaaf931-0dec-11e5-b939-0800200c9a66': {
```
I think that it might make sense to put this as a "device database" in the upstream lib (if possible). That way other consumers of that lib could also use it.

samuellinde commented:
Any updates on this PR?

rdelcampog commented:
> 
> 
> Any updates on this PR?

+1

balloob commented:
This PR seems to have gone stale. Closing it.

You can open a new PR when you have time to finish it.
